lte/gateway/python/magma/pipelined/app/inout.py

```diff
@@ -286,35 +298,37 @@ def _get_gw_mac_address(self, gw_ip: str) -> str:
             self.logger.warning("Error in probing Mac address: err %s", ex)
             return None
 
-    def _monitor_and_update(self, datapath):
-        current_feq = self.config.non_mat_gw_probe_frequency
-        if self._dhcp_gw_info.getMac() is not None:
-            latest_mac_addr = self._dhcp_gw_info.getMac()
-            self._install_default_egress_flows(datapath, latest_mac_addr)
-            flows.set_barrier(datapath)
+    def _monitor_and_update(self):
+        while True:
+            gw_info = get_mobilityd_gw_info()
+            if gw_info:
+                break
+            self.logger.warning("No gw_info")
+            hub.sleep(self.config.non_mat_gw_probe_frequency)
+
+        if gw_info.mac and len(gw_info.mac) == 17:
+            self._install_default_egress_flows(self._datapath, gw_info.mac)
+        else:
+            self.logger.debug("Unexpected MAC: %s", gw_info.mac)
 
         while True:
-            ip = self._dhcp_gw_info.getIP()
-            if ip is not None:
-                self.logger.info("GW found: %s", ip)
-                latest_mac_addr = self._get_gw_mac_address(ip)
-                if latest_mac_addr is not None:
-                    # go back to configured frequency.
-                    current_feq = self.config.non_mat_gw_probe_frequency
+            updated_info = get_mobilityd_gw_info()
+            if updated_info:
+                gw_info = updated_info
+            if gw_info.ip:
+                latest_mac_addr = self._get_gw_mac_address(gw_info.ip)
+                if latest_mac_addr:
                     if self._current_upstream_mac != latest_mac_addr:
-                        self._install_default_egress_flows(datapath, latest_mac_addr)
-                        flows.set_barrier(datapath)
-                        self._dhcp_gw_info.update_mac(latest_mac_addr)
+                        self._install_default_egress_flows(self._datapath, latest_mac_addr)
+                        gw_info.mac = latest_mac_addr
+                    set_mobilityd_gw_info(gw_info)
```

Comment:
You should be able to do this in 1 loop instead of 2. Also you're missing the validation of the mac address on line 309 in this loop. Maybe something along the lines of:

```
current_gw_info = GWInfo()
while True:
    updated_gw_info = get_mobilityd_gw_info()
    if validate(updated_gw_info):
        if gw_info changed:
            _install_egress
        current_... = updated_...
    hub.sleep()
```